### PR TITLE
AOT compilation using ngtools/webpack

### PR DESCRIPTION
**Pull request by deepu105**
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

cc @sendilkumarn 

This is causing some sort of infinite loop in ngc

### CODE REVIEW

**Comment by sendilkumarn** [comment]
infinite loops ?

**Comment by sendilkumarn** [comment]
(check this )[https://stackoverflow.com/questions/26094420/fatal-error-call-and-retry-last-allocation-failed-process-out-of-memory]
and it happens after mergejson is initialized... I dont see any error in my system either.

**Comment by deepu105** [comment]
But this doesnt happen in master only in this branch

File: generators/client/templates/angular/webpack/_webpack.prod.js [hunk 1]
```diff
@@ -38,6 +51,24 @@ module.exports = webpackMerge(commonConfig({ env: ENV }), {
...
+        }),
+        // AOT Plugin 
+        new AotPlugin({
+            tsConfigPath: './tsconfig-aot.json',
+            entryModule: root('src/main/webapp/app/app.module#<%=angular2AppName%>AppModule')
+        }),
+        new webpack.optimize.UglifyJsPlugin({
+            beautify: false,
+            comments: false,
+            sourcMap: true,
```
Comment (deepu105) [inline generators/client/templates/angular/webpack/_webpack.prod.js]: @sendilkumarn is this a typo? shouldnt this be `sourceMap`

**Comment by deepu105** [comment]
@sendilkumarn I found the problem now need to find a fix

**Comment by deepu105** [comment]
@jhipster/developers I'm giving up on this I spend hours on this but couldn't get it working. here is what I did so far
So the idea of using ngtools to do AOT build is so that the generated AOT files go through the same webpack pipeline and hence `ngc` shouldn't be used with ngtools.

But the problem is that somehow its unable to find the ngfactory* files in memory. I tried different samples online https://github.com/blacksonic/angular2-aot-cli-webpack-plugin, https://github.com/seanlandsman/webpack-ngtools-blog etc, and tried to replicate the settings but no luck. I'm unable to figure out what is going wrong. I think we need an angular experts help here @jdubois do you know anyone who could help here?

**Comment by mraible** [comment]
Maybe @TheLarkInn can help?

**Comment by jdubois** [comment]
@deepu105 not sure - the best solution would be to ask for help on Twitter, could you do a message pointing to your explanation here? I will RT it, of course.

**Comment by deepu105** [comment]
@jdubois btw I think this is the only thing blocking Angular out of beta

**Comment by deepu105** [comment]
I did a tweet

**Comment by manekinekko** [comment]
@deepu105 what are the steps to reproduce the issue you're having?

**Comment by deepu105** [comment]
@manekinekko I'll push a repo with sample code now

**Comment by manekinekko** [comment]
great. thanks.

**Comment by deepu105** [comment]
@manekinekko here is the repo https://github.com/deepu105/ngtools-sample
Run the below to reproduce the issue
```
yarn install
yarn webpack:prod
```

The webpack config is under the `./webpack` folder. Angular code is under `./src/main/webapp/app` folder. tsconfig files are at `./`

**Comment by deepu105** [comment]
Hi @manekinekko did you manage to take a look at this?

**Comment by manekinekko** [comment]
Yep, I managed to find some time yesterday evening ^^

The issue was related to the typescript configuration in `tsconfig-aot.json` . 
Here is the config that gets this working:
```
{
    "compilerOptions": {
        "target": "es5",
        "module": "es2015",
        "moduleResolution": "node",
        "sourceMap": true,
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "removeComments": false,
        "noImplicitAny": false,
        "suppressImplicitAnyIndexErrors": true,
        "outDir": "target/www/app",
        "lib": ["es2015", "dom"],
        "typeRoots": [
            "node_modules/@types"
        ]
    },

    "files": [
        "./src/main/webapp/app/app.module.ts",
        "./src/main/webapp/app/app.main-aot.ts"
    ],

    "angularCompilerOptions": {
        "genDir": "target/aot",
        "entryModule": "src/main/webapp/app/app.module#Jh4TestAppModule",
        "skipMetadataEmit": true
    }
}
```
Please note that I'm using the `entryModule` property in the `tsconfig-aot.json`. This is just for convenience. You can leave it in the webpack config.

Also, it seems that webpack has to optimize a lot of files. As a consequence, it requires a lot of memory, so I had to bump up the memory size of node before running webpack:

`node --max_old_space_size=4000 node_modules/webpack/bin/webpack.js -p --config webpack/webpack.prod.js --progress --profile --bail`

And the build process was abnormally long (14min)

image

Here is the devtools report regarding the application bootstrap in AOT vs JIT

## AOT

image

## JIT

image

**Comment by deepu105** [comment]
@manekinekko Thanks a ton for taking the time to look into this. We were also able to get compilation working with something similar but faced the same issue of memory timeout and abnormally long build. Any idea on what would be making the build take such long? Initially, I thought it was some kind of cyclic dependency but couldn't find any. On the other hand compiling just with `ngc` takes around 60 seconds. Would you advice sticking to using `ngc` directly instead of using ngtools?

**Comment by sendilkumarn** [comment]
wow @manekinekko Thanks a lot for taking the time to look into this.

but 14 mins :(  I think we can manually compile and use it rather than using ngtools. one less pain :p

**Comment by deepu105** [comment]
@sendilkumarn if you have time can you clone the branch and try to do it with just `ngc` without ngtools?

**Comment by manekinekko** [comment]
I managed to bring the build down to **2m34s** by disabling the source map in prod mode (`devtool: 'eval'`) and setting `sourceMap: false,` of `webpack.optimize.UglifyJsPlugin` in `webpack.prod.js`. Of course this is not a solution.

One intriguing thing is that processing the html and scss is what takes too long: 
image

@deepu105 The official tool for building Angular in AOT mode is the Angular Compiler which CLI is `ngc`. The webpack plugin uses a private API of the Angular Compiler. So in theory, yes as @sendilkumarn suggested, you can just use `ngc` to build Angular in AOT. But then I guess you'll need to sync the two processes: `ngc` + `webpack` (since you still need to build/optimise the other assets).

Question: I'm interested in hearing why you didn't use the Angular CLI with this generator?

**Comment by sendilkumarn** [comment]
> Question: I'm interested in hearing why you didn't use the Angular CLI with this generator?
@manekinekko infact we tried #5661

**Comment by manekinekko** [comment]
@sendilkumarn I see. sounds fair. I'd argue however that integrating the Angular CLI would make this generator compatible with the official stack.

**Comment by jdubois** [comment]
Maybe we can report the issue on the private API that we use? Otherwise yes we can switch to ngc, but it's always annoying to run external commands as they need to be installed and synced.

**Comment by deepu105** [comment]
Ok, i got this working without using ngtools. I used ngc and then webpack on generated files. It still takes 2 minutes. there are some issue with image paths, i'm gonna try to fix that today

**Comment by deepu105** [comment]
I'm also gonna try to do it like this https://github.com/AngularClass/angular-starter

**Comment by jdubois** [comment]
Great. But those 2 minutes are just for the prod build, right? Just let me state again our priorities, in particular for @manekinekko : 

- In dev mode: very fast startup, with all the tools to debug (source maps, hot reload, BrowserSync)
- In prod mode: minimized files, caching (so this also include our existing work on the server side) -> AOT, tree shaking and Webpack optimizations are the priorities, not the build time

And by the way, doing a full build is going to take several minutes anyway, as we do the Java packaging, all the tests (back end and front end), and then usually build and push a Docker image. So 2 minutes more on this part isn't a big deal.

**Comment by jdubois** [comment]
2 ideas @deepu105 : 

- Would upgrading to Webpack 3 ( #5955 ) improve this? As they claim big performance improvements
- It also look like memory and GC are important, see https://twitter.com/mikesherov/status/881321287364300801

**Comment by deepu105** [comment]
@jdubois I got it working in few different ways, im comparing performance and build time of these and i'll commit the one with best result in this branch soon.

Webpack 3 wont help much here I guess

**Comment by jdubois** [comment]
OK, I'll have time to test tomorrow

**Comment by deepu105** [comment]
@jhipster/developers So finally I got this working with reasonable build times and good performance

**Default app builds in Dev :** `Done in 30.11s.` (12+ MB)
**Default app builds in Prod with AOT :** `Done in 95.82s.` (402KB gzip)
image

An app with lot of entities (23 of our travis sample test entities) : 
**App builds in Dev :** `Done in 50.02.` (14+ MB)
image

**App builds in Prod with AOT :** `Done in 808.20s.` (656KB gzip)
image


Thats around 13 mins for huge app. But as per official docs https://angular.io/guide/aot-compiler#jit-in-development-aot-in-production its seems to be expected

> Today AOT compilation and tree shaking take more time than is practical for development. That will change soon. For now, it's best to JIT compile in development and switch to AOT compilation before deploying to production.
So I guess we may not be able to spped it up further even with Angular CLI. See https://github.com/angular/angular-cli/issues/6516

I used the ngc-webpack plugin for this and thanks to sample from https://github.com/AngularClass/angular-starter

@manekinekko thanks a lot for your inputs and help. WDYT of the implementation I have done?

**Comment by deepu105** [comment]
@sendilkumarn @jdubois please test it

**Comment by deepu105** [comment]
I'm gonna stop 1 of the 2 builds trigerred

**Comment by sendilkumarn** [comment]
Let me have a look. 

@deepu105 but 4xx is slightly less than what we have now right ?

**Comment by sendilkumarn** [comment]
@deepu105 Good Job 👏 

Wow, finally it works but still with ~115 seconds. But as we discussed this earlier, I am fine with that.

**Comment by sendilkumarn** [comment]
there are time outs in build can you check that ? 

screen shot 2017-07-03 at 2 01 55 pm

**Comment by deepu105** [comment]
@sendilkumarn I was expecting that. Those 2 builds have more than 20 entities and they run the prod build. SO build will take more than 15 mins. @pascalgrimaud is it possible to increase the timeout in travis to 20 mins for these builds?

**Comment by deepu105** [comment]
Btw @jdubois what do you think of the build times? is it acceptable? Anyways with AOT I dont think we have any way to speed it up further(Unless there are improvements in Angular/webpack) so the more entities or front end code you have the slower the prod build would be. One more way to reduce the build time by few mins would be to disable sourcemaps in production

**Comment by pascalgrimaud** [comment]
Not sure we can change the timeout on Travis, but I will have a look
Good jobs guys !

**Comment by deepu105** [comment]
@pascalgrimaud I added `travis_wait` to the step as per https://docs.travis-ci.com/user/common-build-problems/#My-builds-are-timing-out

**Comment by pascalgrimaud** [comment]
Oh nice, I didn't know this, thanks @deepu105

**Comment by deepu105** [comment]
@pascalgrimaud btw I also found that Travis has an inbuilt command to retry 3 times on failure (we do that manually for protractor right?)

**Comment by pascalgrimaud** [comment]
exact @deepu105
But I wanted to avoid a retry when there is a real error, only retry for random failures with protractor

**Comment by deepu105** [comment]
ah ok


Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
@deepu105 I'm not sure that sourcemaps in prod are a good idea anyway -> yes can you remove them? Or make them optional, and disabled by default?

I'm really sad to discover **one more time** that Angular does not work well for real apps in production, which was the whole point of switching from AngularJS.

**Comment by deepu105** [comment]
@jdubois source maps are disabled by default. It's commented so people can enable it if required. Maybe we can add a line of documentation regarding it in using webpack/angular page

**Comment by jdubois** [comment]
@deepu105 if I do a prod build of my own sample app (only 3 entities):

- in `app` I still get `app/src` and `app/target` folders, and they take about 5Mb, could you remove them?
- my total build (with Java and all tests) took 2:17 minutes, so that's totally fine for me

**Comment by jdubois** [comment]
@deepu105 -> your build passed, and as far as I'm concerned this all looks OK. I tested against an "old" AngularJS 1 app, and performance isn't great (in fact it's worse!!), but it's basically working OK and I had no issue.

-> let's merge this!!!!! WDYT? I'll let you push the button :-)

**Comment by deepu105** [comment]
@jdubois yes lets merge this. About the folders, are you talking about the build|target folder? if so they are created during AOT compile. I need to see if there is a way to remove it within webpack else after build we need to remove it using `rimraf`

**Pull request merged**